libraries/AP_Param/AP_Param.cpp

```diff
@@ -2662,11 +2676,258 @@ void AP_Param::show_all(AP_HAL::BetterStream *port, bool showKeyValues)
          ap;
          ap=AP_Param::next_scalar(&token, &type)) {
         if (showKeyValues) {
-            ::printf("Key %u: Index %u: GroupElement %u  :  ", (unsigned)_var_info[token.key].key, (unsigned)token.idx, (unsigned)token.group_element);
+            ::printf("Key %u: Index %u: GroupElement %u  :  ", (unsigned)var_info(token.key).key, (unsigned)token.idx, (unsigned)token.group_element);
         }
         show(ap, token, type, port);
         hal.scheduler->delay(1);
     }
 }
 #endif // AP_PARAM_KEY_DUMP
 
+
+#if AP_PARAM_DYNAMIC_ENABLED
+/*
+  allow for dynamically added parameter tables from scripts
+
+  The layout we create is as follows:
+    - a top level Info with the given prefix, using one of the 10 possible slots in _var_info_dynamic
+    - a dynamically allocated GroupInfo table, never freed, of size (num_params+2)
+    - the GroupInfo table has an initial AP_Int32 hidden entry with a 32 bit CRC of the prefix
+    - the last GroupInfo is a footer
+*/
+bool AP_Param::add_table(uint8_t _key, const char *prefix, uint8_t num_params)
+{
+    // check if the key already exists. We only check base params to allow
+    // for scripting reload without a conflict
+    uint16_t key = uint16_t(_key) + AP_PARAM_DYNAMIC_KEY_BASE;
+    for (uint16_t i=0; i<_num_vars_base; i++) {
+        if (var_info(i).key == key) {
+            return false;
+        }
+    }
+    if (num_params > 63) {
+        return false;
+    }
+
+    // we use a crc of the prefix to ensure the table key isn't re-used
+    const int32_t crc = int32_t(crc32_small(0, (const uint8_t *)prefix, strlen(prefix)));
+    int32_t current_crc;
+    if (load_int32(key, 0, current_crc) && current_crc != crc) {
+        // crc mismatch, we have a conflict with an existing use of this key
+        return false;
+    }
+
+    // create the dynamic table if needed. This is never freed
+    if (_var_info_dynamic == nullptr) {
+        _var_info_dynamic = (Info *)calloc(AP_PARAM_MAX_DYNAMIC, sizeof(struct Info));
+        if (_var_info_dynamic == nullptr) {
+            return false;
+        }
+        for (uint8_t i=0; i<AP_PARAM_MAX_DYNAMIC; i++) {
+            auto &info = _var_info_dynamic[i];
+            info.type = AP_PARAM_NONE;
+            info.name = _empty_string;
+            info.key = 0xFFFF;
+            info.ptr = nullptr;
+            info.group_info = nullptr;
+            info.flags = 0;
+        }
+        // make tables available
+        _num_vars += AP_PARAM_MAX_DYNAMIC;
+    }
+
+    // find existing key (allows for script reload)
+    uint8_t i;
+    for (i=0; i<AP_PARAM_MAX_DYNAMIC; i++) {
+        auto &info = _var_info_dynamic[i];
+        if (info.type != AP_PARAM_NONE && info.key == key) {
+            if (_dynamic_table_sizes[i] != 0 &&
+                num_params > _dynamic_table_sizes[i]) {
+                // can't expand the table at runtime
+                return false;
+            }
+            if (strcmp(prefix, info.name) != 0) {
```

Comment:
This case is quite unlikely now we have the CRC check

Reply:
protects against hash collisions (eg. short prefix)